script to automate updates to robots.txt

cxxboy opened the pull request:
Hi, I'm first time contributing to solidity.
I'm trying to solve #11199 

I have written a Perl script which reads the `robots.txt` and checks the currently allowed version against the one returned by `get_version.sh` and if it sees the same version numbers(i.e there is no changes from the output of `get_version.sh` and the one allowed on `robots.txt`) the script does nothing and exits. If it sees the different version number from the one outputted by `get_version.sh`, it changes the version and makes the previous version as `Disallow`

Example, if the `robots.txt` file have 
```
...
# Allow the latest release
#Disallow: /en/v0.8.3/
```  
but the version returned by `get_version.sh` is `0.8.4`, it changes the `robots.txt` file as follows

```
...
# Allow the latest release
Disallow: /en/v0.8.3/
#Disallow: /en/v0.8.4/
```
But if the `get_version.sh` script also outputs `0.8.3`, `robots.txt` file remains untouched.

Let me know if there are any comments, suggestions or changes.

cameel commented:
Hello @cxxboy. Thanks for contributing!

The thing is, #10972 has not been merged yet so I think it would be best to wait for a conclusion there. Looks like I might have been too quick with creating an issue for the autoupdate but I did not think someone would submit a PR for that already :)

The other thing is that we already have scripts written in a bunch of languages (Bash, JavaScript, Python) and I think that adding another one to the mix might not be the best idea. Honestly, I don't even know Perl. Not sure about others on the team. fI see that the script is invoking Bash commands like `sed` anyway so apparently it's not for portability, so why not use Bash?

cxxboy commented:
@cameel 
Sorry about late reply, I have pushed a bash version like you suggested.
But not sure why the CI test fails.

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+#
+#     https://docs.soliditylang.org
+#
+# ------------------------------------------------------------------------------
```
This description does not match the script.

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2017 solidity contributors.
```
In newly created files please use the current year:
```suggestion
# (c) 2021 solidity contributors.
```

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+#------------------------------------------------------------------------------
+
+set -e
+set -u
```
You can shorten this:
```suggestion
set -eu
```

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+set -e
+set -u
+
+PATH_TO_GET_VERSION_SCRIPT="./scripts/get_version.sh"
+PATH_TO_ROBOTS_TXT_FILE="docs/_static/robots.txt"
```
The script cannot find this path when running in CI. Here's [the error from the `b_docs` CI job](https://app.circleci.com/pipelines/github/ethereum/solidity/14768/workflows/17530589-877e-4e39-95d8-2fd4e075dc4b/jobs/675068):
```
tail: cannot open 'docs/_static/robots.txt' for reading: No such file or directory
```

It's probably because you used a relative path and the script is not being executed from repository root. Please change it so that the script works no matter where it's executed from.

See how we're doing it in other scripts: https://github.com/ethereum/solidity/blob/f895760bbcde3b32e50b355e15d230fcc06950b6/scripts/tests.sh#L31

cameel replied:
The typical Bash convention is to use UPPERCASE names only for environment and internal shell variables. Other variables should be lowercase.

Also, I'd make these names a bit more succint:

```suggestion
get_version_script_path="./scripts/get_version.sh"
robots_txt_path="docs/_static/robots.txt"
```

BTW, I know that our existing scripts do not use this convention consistently but we should stick to it at least when writing new ones.

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+
+PATH_TO_GET_VERSION_SCRIPT="./scripts/get_version.sh"
+PATH_TO_ROBOTS_TXT_FILE="docs/_static/robots.txt"
+TEMP_FILE="temp_store_robots.txt"
```
It's not safe to assume that you have write permission in the current working directory. You should create a file in the system tmp dir instead. See how we're doing it in other scripts:
https://github.com/ethereum/solidity/blob/f895760bbcde3b32e50b355e15d230fcc06950b6/scripts/check_style.sh#L5

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+PATH_TO_ROBOTS_TXT_FILE="docs/_static/robots.txt"
+TEMP_FILE="temp_store_robots.txt"
+VERSION_FROM_SCRIPT=`bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp'`
+LAST_LINE_ROBOTS=`tail -n 1 ${PATH_TO_ROBOTS_TXT_FILE}`
```
This is one of the things that [shellcheck](https://www.shellcheck.net/) is [complaining about in `chk_coding_style` job in CI](https://app.circleci.com/pipelines/github/ethereum/solidity/14768/workflows/17530589-877e-4e39-95d8-2fd4e075dc4b/jobs/675064):

```

In ./scripts/update_robots.sh line 35:
VERSION_FROM_SCRIPT=`bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp'`
                    ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean: 
VERSION_FROM_SCRIPT=$(bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp')


In ./scripts/update_robots.sh line 36:
LAST_LINE_ROBOTS=`tail -n 1 ${PATH_TO_ROBOTS_TXT_FILE}`
                 ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean: 
LAST_LINE_ROBOTS=$(tail -n 1 ${PATH_TO_ROBOTS_TXT_FILE})
```

```suggestion
VERSION_FROM_SCRIPT=$(bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp')
LAST_LINE_ROBOTS=$(tail -n 1 ${PATH_TO_ROBOTS_TXT_FILE})
```

By the way, you can run the same checks locally with `scripts/chk_shellscripts/chk_shellscripts.sh` from repo root.

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+PATH_TO_GET_VERSION_SCRIPT="./scripts/get_version.sh"
+PATH_TO_ROBOTS_TXT_FILE="docs/_static/robots.txt"
+TEMP_FILE="temp_store_robots.txt"
+VERSION_FROM_SCRIPT=`bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp'`
```
You don't need to explicitly run `bash` to execute a script.

Also, I don't think the variable is necessary. It's more readable to use script name directly.

Finally, I think the name should say what this version refers to, not where it comes from.

```suggestion
solidity_version=`./scripts/get_version.sh | perl -pe 'chomp'`
```

cameel replied:
I don't think adding a dependency on `perl` just to strip the trailing newline is a good idea. Can you replace with something else? E.g. `tr` or a regex?

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+[[ ${LAST_LINE_ROBOTS} =~ ${VERSION_NUMBER_IN_ROBOTS} ]]
+if [[ $? == 0 ]]; then
+    if [[ ${BASH_REMATCH[2]} != ${VERSION_FROM_SCRIPT} ]]; then
+        sed '$d' ${PATH_TO_ROBOTS_TXT_FILE} > ${TEMP_FILE}; cat ${TEMP_FILE} > ${PATH_TO_ROBOTS_TXT_FILE}
```
It's more readable with one command per line in my opinion:
```suggestion
        sed '$d' ${PATH_TO_ROBOTS_TXT_FILE} > ${TEMP_FILE}
        cat ${TEMP_FILE} > ${PATH_TO_ROBOTS_TXT_FILE}
```

cameel replied:
You're copying a file. Why use `cat` rather than `cp`?

cameel replied:
`sed` can modify a file in place.

Also, I think this needs a comment because I don't think that the `$d` sed command is something the everyone will immediately recognize.

```suggestion
        # Remove the last line from robots.txt
        sed -i.bak '$d' "${PATH_TO_ROBOTS_TXT_FILE}"
```

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+TEMP_FILE="temp_store_robots.txt"
+VERSION_FROM_SCRIPT=`bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp'`
+LAST_LINE_ROBOTS=`tail -n 1 ${PATH_TO_ROBOTS_TXT_FILE}`
+VERSION_NUMBER_IN_ROBOTS="#\w+:\s\/([a-z]+)\/v([0-9]\.[0-9]\.[0-9])\/"
+[[ ${LAST_LINE_ROBOTS} =~ ${VERSION_NUMBER_IN_ROBOTS} ]]
+if [[ $? == 0 ]]; then
```
Another warning from [`chk_coding_style` job in CI](https://app.circleci.com/pipelines/github/ethereum/solidity/14768/workflows/17530589-877e-4e39-95d8-2fd4e075dc4b/jobs/675064):
```
In ./scripts/update_robots.sh line 39:
if [[ $? == 0 ]]; then
      ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.
```

This is way more readable on a single line and that form does not require you to check the exit code indirectly:

```suggestion
if [[ $LAST_LINE_ROBOTS =~ #\w+:\s\/([a-z]+)\/v([0-9]\.[0-9]\.[0-9])\/ ]]; then
```

cameel replied:
If the last line does not match the pattern, e.g. is not a `Disallow` directive, the script just ignores the problem and exits without saying anything. I don't think this is the right behavior. The script should print an error message and exit in that case so that we can immediately spot the problem and fix it.

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+VERSION_NUMBER_IN_ROBOTS="#\w+:\s\/([a-z]+)\/v([0-9]\.[0-9]\.[0-9])\/"
+[[ ${LAST_LINE_ROBOTS} =~ ${VERSION_NUMBER_IN_ROBOTS} ]]
+if [[ $? == 0 ]]; then
+    if [[ ${BASH_REMATCH[2]} != ${VERSION_FROM_SCRIPT} ]]; then
```
Last warning from [`chk_coding_style` job in CI](https://app.circleci.com/pipelines/github/ethereum/solidity/14768/workflows/17530589-877e-4e39-95d8-2fd4e075dc4b/jobs/675064):

```
In ./scripts/update_robots.sh line 40:
    if [[ ${BASH_REMATCH[2]} != ${VERSION_FROM_SCRIPT} ]]; then
                                ^--------------------^ SC2053: Quote the right-hand side of != in [[ ]] to prevent glob matching.
```

You're not quoting variables properly.

Also, you do not need the `${}` around every variable. It's only needed when the variable is surrounded by text, e.g. `echo "XXX${value}XXX"`. I generally use it when I need to use a variable somewhere inside a bigger string but I think that it just adds noise when the variable stands on its own.

```suggestion
    if [[ $BASH_REMATCH[2] != "$VERSION_FROM_SCRIPT" ]]; then
```

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+[[ ${LAST_LINE_ROBOTS} =~ ${VERSION_NUMBER_IN_ROBOTS} ]]
+if [[ $? == 0 ]]; then
+    if [[ ${BASH_REMATCH[2]} != ${VERSION_FROM_SCRIPT} ]]; then
+        sed '$d' ${PATH_TO_ROBOTS_TXT_FILE} > ${TEMP_FILE}; cat ${TEMP_FILE} > ${PATH_TO_ROBOTS_TXT_FILE}
+        echo "Disallow: /${BASH_REMATCH[1]}/v${BASH_REMATCH[2]}/" >> ${PATH_TO_ROBOTS_TXT_FILE}
+        echo "#Disallow: /${BASH_REMATCH[1]}/v${VERSION_FROM_SCRIPT}/" >> ${PATH_TO_ROBOTS_TXT_FILE}
```
The script makes an assumption that the line we want to change is the last one in the file. I don't think this is a good assumption. We're likely to add other things to the file later and the script should handle them gracefully. It would be best if we don't have to update this script every time someone changes something in `robots.txt`.

The current structure of the file is not well suited for detecting the right `Disallow` line to change so my proposal would be to change its structure a bit. First let's manually update `robots.txt` to replace this:

```robots.txt
# Allow the latest release
#Disallow: /en/v0.8.3/
```

with this:
```robots.txt
# LATEST RELEASE: 0.8.3
```

Then, when version in `LATEST RELEASE` does not match, have the script add a new `Disallow` line before it and then update the version on the `LATEST RELEASE` line.

I'd also add some sanity checks:
- Exit with an error if a line of the form `# LATEST RELEASE: x.y.z` is not present or if there is more than one such line
- Exit with an error if there already is a `Disallow` directive that has the same version as the one used on the `LATEST RELEASE` line.

cxxboy replied:
@cameel 
So, can I implement in a way that the script will look for the "last" `#Disallow` and `Disallow` and tries to modify that? Assuming we will add some more stuff onto the `robots.txt`, I think by this way we don't rely on changing the final line right?

cameel replied:
I think that just looking at `Disallow` is not realiable enough. See for example how the current script badly updated `robots.txt` in the new PR.

I think that the system I suggested above would be more robust.

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+TEMP_FILE="temp_store_robots.txt"
+VERSION_FROM_SCRIPT=`bash ${PATH_TO_GET_VERSION_SCRIPT} | perl -pe 'chomp'`
+LAST_LINE_ROBOTS=`tail -n 1 ${PATH_TO_ROBOTS_TXT_FILE}`
+VERSION_NUMBER_IN_ROBOTS="#\w+:\s\/([a-z]+)\/v([0-9]\.[0-9]\.[0-9])\/"
```
For now I'd assume that the path is `/en/vx.y.z/`. If we add more prefixeds we'll have to adjust the script anyway:
```suggestion
VERSION_NUMBER_IN_ROBOTS="#\w+:\s\/en\/v([0-9]\.[0-9]\.[0-9])\/"
```

cameel commented on scripts/docs.sh:
```diff
@@ -31,3 +31,4 @@ cd docs
 pip3 install -r requirements.txt
 sphinx-build -nW -b html -d _build/doctrees . _build/html
 cd ..
+bash scripts/update_robots.sh
```
```suggestion
scripts/update_robots.sh
```

cameel commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
```suggestion
#!/usr/bin/env bash
```

cameel reviewed (changes requested):
The script needs some adjustments.

> But not sure why the CI test fails(since I cannot see the CI's logs)

It's because your script triggers warnings from shellcheck. I added comments with fix suggestions below.

You are also using relative paths and that does not work in CI because the script is not executed from repo root. You need to use paths relative to script location.

cameel commented:
One more thing: the last point on [the release checklist for docs](https://github.com/ethereum/solidity/blob/2962d48593fb5542af7049d891777ecd896addaf/ReleaseChecklist.md#documentation) should be updated.

cameel commented:
@cxxboy How is the progress on the PR? Are you still working on it?

We're having a cleanup week right now and trying to reduce the number of pending PRs (we've hit 60+ at one point). Is there a way I could help you finish it and get it merged?

cxxboy reviewed (commented):
> @cxxboy How is the progress on the PR? Are you still working on it?
> 
> We're having a cleanup week right now and trying to reduce the number of pending PRs (we've hit 60+ at one point). Is there a way I could help you finish it and get it merged?

@cameel 
The major changes you requested is done, I was wondering why the `b_docs` can't find the `robots.txt` from the project-root, but found that my branch doesn't have the file so I fixed it. Regarding the last comment, we probably need to iterate from the file's end like Last-to-First to find the first `#Disallow` string and match against that but as of now, the script assumes that we need to change the last line. Other than that, everything is updated as per your comments. 

But I can see some other CI test is failing but it's not related to my code, so I can't figure out why that fails and how to fix that.

chriseth commented on scripts/update_robots.sh:
```diff
@@ -0,0 +1,38 @@
+
+repo_root="$(dirname "$0")/.."
+robots_txt_path="${repo_root}/docs/_static/robots.txt"
+version_from_script=$("${repo_root}"/scripts/get_version.sh | perl -pe 'chomp')
```
Please don't add perl as a dependency.

chriseth replied:
I don't get any unnecessary whitespace from `get_version.sh` - maybe you can just use
```suggestion
version_from_script=$("${repo_root}"/scripts/get_version.sh)
```
here?

cxxboy replied:
Removed Perl and used `tr`. It's for removing the trailing new line from `get_version.sh`

chriseth replied:
Are you sure it is needed? I think the shell already removes the trailing newline on its own...

chriseth reviewed (commented):
Thanks a lot for updating!
It looks good apart from two items:
 - the perl command
 - please rebase on top of develop and remove the merge commit

cxxboy committed:
script to automate updates to robots.txt #11199

The bash scripts reads the docs/_static/robots.txt file and checks the
version against the output of 'get_version.sh' script and updates the
robots.txt if the version doesn't match with the one on the robots.txt
file. If it finds the same version, the scripts does nothing(robots.txt
remains unmodified since the version matched)
Since it's executed as part of 'scripts/docs.sh', I added the script in
there.

cxxboy commented:
> Thanks a lot for updating!
> It looks good apart from two items:
> 
>     * the perl command
> 
>     * please rebase on top of develop and remove the merge commit

Hi @chriseth, I'm very new to rebase workflow. I had followed some instructions and rebased my branch on-top-of develop. And I had to resolve a conflict on 'robots.txt' from the other person during rebase. But looks like on my commit, it added whole bunch of stuff: https://github.com/cxxboy/solidity/commit/67b59246ba528f5aee614aa1e77467e888890556
Can you please help me with this one?

cameel commented:
@cxxboy I think that the problem is that your code is still not based on latest `develop`. Could you make sure that you're rebasing on on `dce3006723e56f60c7f89d6095e14298ef154945`?

BTW, I recommend interactive rebase so that you can easily pick and drop commits.

First make sure your `develop` branch is up to date:
```bash
git fetch origin
git check develop
git merge origin/develop --ff-only
```
Then rebase:
```bash
git checkout update_robots_script
git rebase develop --interactive
```

The `--interactive` option will open an editor where you can select commits to be included. Make sure you pick only your own commits. Any merge commits or base commits carried over from previous location should be dropped.

cxxboy commented:
> @cxxboy I think that the problem is that your code is still not based on latest `develop`. Could you make sure that you're rebasing on on `dce3006723e56f60c7f89d6095e14298ef154945`?
> 
> BTW, I recommend interactive rebase so that you can easily pick and drop commits.
> 
> First make sure your `develop` branch is up to date:
> 
> ```shell
> git fetch origin
> git check develop
> git merge origin/develop --ff-only
> ```
> 
> Then rebase:
> 
> ```shell
> git checkout update_robots_script
> git rebase develop --interactive
> ```
> 
> The `--interactive` option will open an editor where you can select commits to be included. Make sure you pick only your own commits. Any merge commits or base commits carried over from previous location should be dropped.

I'm really sorry, @cameel I think I messed something up real bad so I think I need to start over again. I cannot seem to recover. Can I close this PR and make a new one with the updated scripts in there?

cameel commented:
Well, we still have the commit IDs of previous versions you pushed so it should be easy to recover. I'll post the commands in a moment.

In the meantime, I tried to find some articles that would explain the workflow nicely. Maybe you could take a look at one of these?
- https://stackoverflow.com/questions/804115/when-do-you-use-git-rebase-instead-of-git-merge/46708899#46708899
- https://github.com/thewoolleyman/git-rebase-workflow-example1/blob/master/README.md
- https://medium.com/singlestone/a-git-workflow-using-rebase-1b1210de83e5

cameel commented:
I see that the last commit without the mess is 5b74e9e290ef83a731fe9179dc7ccadc0d90a2a7. Does it contain all your changes? If so, you can move your branch back to that place with:

```bash
# Make a backup label at the current location of your branch, just in case
git checkout update_robots_script
git checkout -B backup-update_robots_script

# Move the branch to an old commit
git checkout 5b74e9e290ef83a731fe9179dc7ccadc0d90a2a7
git checkout -B update_robots_script
```

cxxboy commented:
@cameel The commit 5b74e9e290ef83a731fe9179dc7ccadc0d90a2a7, doesn't have one change that I made to remove Perl and used `tr`.

cameel commented:
Ok. We we can try differently. Do you have a copy of the latest version of the file you modified? If so you can just reset the branch and recreate the commits from scratch:

```
git checkout develop
git checkout -B update_robots_script
```

**NOTE: This will remove all changes from your branch and move it to the current `develop`**. Then you will be able to replace the file with the copy, commit it and push the branch.

cameel commented:
Just to be clear: I think it would be fine to start a new PR but I think it's not really needed because fixing this one should be pretty simple. If you do https://github.com/ethereum/solidity/pull/11206#issuecomment-826871612, you basically start from scratch with your branch anyway. git is nice in that it does not really modify your commits, it always creates new ones so it's possible to go back to a previous state from almost any situation.

cxxboy commented:
@cameel The commit 67b59246ba528f5aee614aa1e77467e888890556 sticks with the develop so even if I follow your instruction, I end up with that messy state. I think it's easy for me if I close this PR and start a fresh one with the updated scripts.
And regarding that, I will be cloning the fresh repo and make the changes in `robots_update_script` branch and follow(and previously followed) these commands to rebase the develop. But I'm not sure if this is what I need to do:
```
git checkout robots_update_script
// I makes the changes
git commit
git rebase develop
git checkout develop
git rebase robots_update_script
git push origin robots_update_script
// and I open a new PR
```
Can you please see if these steps look good and I can just follow this?

cameel commented:
No, this is not correct. Especially this part makes no sense:
```bash
git checkout develop
git rebase robots_update_script
```

I think you're trying to use `rebase` like `merge`. You're first rebasing your branch on `develop` but then turning around and trying to rebase `develop` on it. Which is not an error but is totally not what you want. This just moves `develop` label to point at the same commit as your branch. This is why that commit "sticks".

You should reset your `develop` to match the upstream.

Here's how it should work:

```bash
# To make sure your develop matches the one from the repository:
git fetch origin
git checkout origin/develop
git checkout develop

git checkout robots_update_script
# Make the changes
git commit

# Always use interactive rebase. When the editor opens, ensure that the list contains only your own
# commits and no older commits are getting dragged along with it. If they are, just remove them from
# the list
git rebase develop --interactive

# Force push is usually needed after you rebase. This is fine as long as you're only doing it with your
# feature branch (and not e.g. `develop`).
git push origin robots_update_script --force-with-lease

# If you do all the above, there's no need to open a new PR.
# When you push, this one will be updated just fine.
```

chriseth commented:
Here is a shorter version that does not use the local `develop` branch:
```
# Ensure that your develop matches the one from the repository
git fetch origin
git checkout origin/develop

git checkout robots_update_script
# Make the changes
git commit

# Always use interactive rebase. When the editor opens, ensure that the list contains only your own
# commits and no older commits are getting dragged along with it. If they are, just remove them from
# the list
git rebase origin/develop --interactive

# Force push is usually needed after you rebase. This is fine as long as you're only doing it with your
# feature branch (and not e.g. `develop`).
git push origin robots_update_script --force-with-lease

# If you do all the above, there's no need to open a new PR.
# When you push, this one will be updated just fine.
```

cameel commented:
Yeah. Though if the local branch has diverged I think it's better if @cxxboy fixes it too :)

cxxboy commented:
@chriseth @cameel I end up starting from fresh. My new PR is #11317
